contribute/architecture/state.md

```diff
@@ -0,0 +1,142 @@
+
+> **Deprecated:** Refer to the [deprecation note for `sqlstore` handlers](#sqlstore-handlers).
+
+To register a handler:
+
+- Create a new file `myrepo.go` in the `sqlstore` package.
+- Create a [command handler](services.md#handle-commands).
+- Register the handler in the `init` function:
+
+```go
+func init() {
+    bus.AddHandler("sql", DeleteDashboard)
+}
+
+func DeleteDashboard(cmd *models.DeleteDashboardCommand) error {
+    return inTransaction(func(sess *DBSession) error {
+        _, err := sess.Exec("DELETE FROM dashboards WHERE dashboard_id=?", cmd.DashboardID)
+        return err
+    })
+}
+```
+
+Here, `inTransaction` is a helper function in the `sqlstore` package that provides a session, that lets you execute SQL statements.
+
+## `SqlStore`
+
+As opposed to a `sqlstore` handler, the `SqlStore` is a service itself. The `SqlStore` has the same responsibility however: to store and retrieve objects, to and from the database.
+
+To use the `SqlStore`, inject the `SQLStore` in your service struct:
+
+```go
+type MyService struct {
+    SQLStore *sqlstore.SqlStore `inject:""`
+}
+```
+
+You can now make SQL queries in any of your [command handlers](services.md#handle-commands) or [event listeners](services.md#subscribe-to-events):
+
+```go
+func (s *MyService) DeleteDashboard(cmd *models.DeleteDashboardCommand) error {
+    if err := s.SQLStore.WithDbSession(ctx, func(sess *sqlstore.DBSession) error {
+        _, err := sess.Exec("DELETE FROM dashboards WHERE dashboard_id=?", cmd.DashboardID)
+        return err
+    })
+}
+```
+
+For transactions, use the `WithTransactionalDbSession` method instead.
+
+## Migrations
+
+As Grafana evolves, it becomes necessary to create _schema migrations_ for one or more database tables.
+
+To see all the types of migrations you can add, refer to [migrations.go](/pkg/services/sqlstore/migrator/migrations.go).
+
+Before you add a migration, make sure that you:
+
+- Never change a migration that has been committed and pushed to master.
+- Always add new migrations, to change or undo previous migrations.
+
+Add a migration using one of the following methods:
+
+- Add migrations in the `migrations` package.
+- Implement the `DatabaseMigrator` for the service.
+
+**Important:** If there are previous migrations for a service, use that method. By adding migrations using both methods, you risk running migrations in the wrong order.
+
+### Add migrations in `migrations` package
+
+Most services have their migrations located in the [migrations](https://github.com/grafana/grafana/blob/master/pkg/services/sqlstore/migrations/migrations.go) package.
+
+To add a migration:
+
+- Open the [migrations.go](https://github.com/grafana/grafana/blob/master/pkg/services/sqlstore/migrations/migrations.go) file.
+- In the `AddMigrations` function, find the `addXxxMigration` function for the service you want to create a migration for.
+- At the end of the `addXxxMigration` function, register your migration:
+
+```go
```

Comment:
Code examples tends to become out of date quite easily. Maybe consider linking to a permalink of code or maybe markdown files don't support rendering it same way as GitHub comments does it? For example https://github.com/grafana/grafana/blob/00d0640b6e778ddaca021670fe851fe00982acf2/pkg/services/sqlstore/migrations/migrations.go#L55-L70